src/compiler/checker.ts

```diff
@@ -14531,16 +14531,12 @@ namespace ts {
              * if we have found an elaboration, or we should ignore
              * any other elaborations when relating the `source` and
              * `target` types.
-             *
-             * @param source
-             * @param target
-             * @param reportErrors
              */
             function tryElaborateArrayLikeErrors(source: Type, target: Type, reportErrors: boolean): boolean {
                 if (isTupleLikeType(source)) {
                     const sourceTuple: TupleType | undefined  = (source as TupleTypeReference).target;
                     if (sourceTuple && sourceTuple.readonly && isArrayOrTupleLikeType(target) &&
-                        (!isReadonlyArrayType(target) || isTupleType(target) && !target.target.readonly)) {
+                        !isReadonlyArrayType(target) && !(isTupleLikeType(target) && (target as TupleTypeReference).target.readonly)) {

```

Comment:
A tuple-like type just needs to have a `0` property, so this cast isn't safe.

Reply:
Alternative is just to dumb down the error message and check if it's strictly a tuple type or array type

Reply:
Sorry, was abit confused by the comment on `isTupleLikeType`.

Reply:
Does this check give the right error message for the following?

```ts
interface MyTupleLike { readonly 0: string }

let x: MyTupleLike = [10000] as const;
```

Reply:
To elaborate, I think we're getting close, but

```ts
isArrayOrTupleLikeType(target)
```

probably needs to be replaced with something like

```ts
(isArrayType(target) || isTupleType(target))
```

so get rid of the `Like`

Reply:
The error message is now:
```ts
Type 'readonly [10000]' is not assignable to type 'MyTupleLike'.
  Types of property '0' are incompatible.
    Type '10000' is not assignable to type 'string'. [2322]
```

Reply:
Some other baseline changes that remove: `... pop, push, concat, and 16 more.`. Is that reasonable?